addons/godot_rl_agents/sync.gd

```diff
@@ -2,165 +2,253 @@ extends Node
+	get_tree().set_pause(true)
 	_initialize()
 	await get_tree().create_timer(1.0).timeout
-	get_tree().set_pause(false) 
-	
+	get_tree().set_pause(false)
+
+
 func _initialize():
 	_get_agents()
-	_obs_space = agents[0].get_obs_space()
-	_action_space = agents[0].get_action_space()
 	args = _get_args()
-	Engine.physics_ticks_per_second = _get_speedup() * 60 # Replace with function body.
+	Engine.physics_ticks_per_second = _get_speedup() * 60  # Replace with function body.
 	Engine.time_scale = _get_speedup() * 1.0
-	prints("physics ticks", Engine.physics_ticks_per_second, Engine.time_scale, _get_speedup(), speed_up)
-	
-	_set_heuristic("human")
-	match control_mode:
-		ControlModes.TRAINING:
-			connected = connect_to_server()
-			if connected:
-				_set_heuristic("model")
-				_handshake()
-				_send_env_info()  
-			else:
-				push_warning("Couldn't connect to Python server, using human controls instead. ",
-				"Did you start the training server using e.g. `gdrl` from the console?")
-		ControlModes.ONNX_INFERENCE:
-				assert(FileAccess.file_exists(onnx_model_path), "Onnx Model Path set on Sync node does not exist: %s" % onnx_model_path)
-				onnx_model = ONNXModel.new(onnx_model_path, 1)
-				_set_heuristic("model")	
-	
+	prints(
+		"physics ticks",
+		Engine.physics_ticks_per_second,
+		Engine.time_scale,
+		_get_speedup(),
+		speed_up
+	)
+
+	_set_heuristic("human", all_agents)
+
+	_initialize_training_agents()
+	_initialize_inference_agents()
+
 	_set_seed()
 	_set_action_repeat()
-	initialized = true  
-
-func _physics_process(_delta): 
+	initialized = true
+
+func _initialize_training_agents():
+	if agents_training.size() > 0:
+		_obs_space = agents_training[0].get_obs_space()
+		_action_space = agents_training[0].get_action_space()
+		connected = connect_to_server()
+		if connected:
+			_set_heuristic("model", agents_training)
+			_handshake()
+			_send_env_info()
+		else:
+			push_warning(
+				"Couldn't connect to Python server, using human controls instead. ",
+				"Did you start the training server using e.g. `gdrl` from the console?"
+			)
+
+func _initialize_inference_agents():
+	if agents_inference.size() > 0:
+		if control_mode == ControlModes.ONNX_INFERENCE:
+			assert(
+				FileAccess.file_exists(onnx_model_path),
+				"Onnx Model Path set on Sync node does not exist: %s" % onnx_model_path
+			)
+			onnx_models[onnx_model_path] = ONNXModel.new(onnx_model_path, 1)
+
+		for agent in agents_inference:
+			_action_space_inference.append(agent.get_action_space())
+
+			var agent_onnx_model: ONNXModel
+			if agent.onnx_model_path.is_empty():
+				assert(
+					onnx_models.has(onnx_model_path),
+					(
+						"Node %s has no onnx model path set " % agent.get_path()
+						+ "and sync node's control mode is not set to OnnxInference. "
+						+ "Either add the path to the AIController, "
+						+ "or if you want to use the path set on sync node instead, "
+						+ "set control mode to OnnxInference."
+					)
+				)
+				prints(
```

Comment:
This info is printed even if AIController is set to inherit mode, but has no onnx model path (then onnx model path from sync node is used).

I'm still thinking about the optimal way to decide when to use which path. Right now you can override the mode, and/or you can override the path in AIController.